docs/protocol-methods.rst

```diff
@@ -498,6 +568,119 @@ Unsubscribe from a script hash, preventing future notifications if its :ref:`sta
   Note that :const:`False` might be returned even for something subscribed to earlier,
   because the server can drop subscriptions in rare circumstances.
 
+blockchain.outpoint.subscribe
+=============================
+
+Subscribe to a transaction outpoint (TXO), to get notifications about its status.
+A status involves up to two transactions: the funding transaction that creates
+the TXO (as one of its outputs), and the spending transaction that uses it
+as an input (spends it).
+
+**Signature**
+
+  .. function:: blockchain.outpoint.subscribe(tx_hash, txout_idx, spk_hint=None)
+  .. versionadded:: 1.5
+
+  *tx_hash*
+
+    The TXID of the funding transaction as a hexadecimal string.
+    (sometimes called prevout_hash, in inputs)
+
+  *txout_idx*
+
+    The output index, a non-negative integer. (sometimes called prevout_n, in inputs)
+
+  *spk_hint*
+
+    The scriptPubKey (output script) corresponding to the outpoint, as a hexadecimal
+    string. This is optional, and if provided might be used by the server to find
+    the outpoint. The behaviour is undefined if an incorrect value is provided.
+    Note that some implementations (such as EPS/BWT) might require this parameter
+    to be able to serve the request.
+
+**Result**
+
+  The status of the TXO, taking the mempool into consideration.
+  The output is a dictionary, containing 0, 1, or 3 of the following items:
+
+  * *height*
+
+    The integer height of the block the funding transaction was confirmed in.
+    If the funding transaction is in the mempool; the value is
+    ``0`` if all its inputs are confirmed, and ``-1`` otherwise.
+    This key must be present if and only if there exists a funding transaction
+    (either in the best chain or in the mempool), regardless of spentness.
+
+  * *spender_txhash*
+
+    The TXID of the spending transaction as a hexadecimal string.
+    This key is present if and only if there exists a spending transaction
+    (either in the best chain or in the mempool).
+
+  * *spender_height*
+
+    The integer height of the block the spending transaction was confirmed in.
+    If the spending transaction is in the mempool; the value is
+    ``0`` if all its inputs are confirmed, and ``-1`` otherwise.
+    This key is present if and only if the *spender_txhash* key is present.
+
+
+**Result Examples**
+
+::
+
+  {}
+
+::
+
+  {
+    "height": 1866594
+  }
+
+::
+
+  {
+    "height": 1866594,
+    "spender_txhash": "4a19a360f71814c566977114c49ccfeb8a7e4719eda26cee27fa504f3f02ca09",
+    "spender_height": 0
+  }
+
+**Notifications**
+
+  The client will receive a notification when the `status` of the outpoint changes.
```

Comment:
Does this meant that the client will receive a notification for each of these events?

* All inputs of the funding transaction get confirmed
* The funding transaction gets confirmed
* The spender appears in the mempool
* The spender is confirmed

So potentially four notifications (but possibly less if everything is confirmend in a single block).

Reply:
> 0. funding tx does not exist

Oh, I would expect this to avoid subscription. I think this should be called out explicitly (and perhaps other states too).

`-1` -> `0` is probably fine even though I guess it'll put additional load on the server.

Reply:
Actually, I had forgotten about reorgs: with reorgs there is no upper bound on the max number of notifications. A single reorg that only affects one subscribed outpoint can result in up to 3 additional notifications over the lifecycle of the outpoint: a reorg can turn a mined tx to height -1 (1st notif), which then can turn into height 0 in a later block (2nd notif), and then get mined in a later block (3rd notif).

> The client will receive a notification when the `status` of the outpoint changes.

This is why the highlighted sentence is written as such: I wanted a generic phrasing that reflects that any time the contents of the status dict change, the server is supposed to send a notification including the new status to the client.

Reply:
>> 0. funding tx does not exist
>
> Oh, I would expect this to avoid subscription. I think this should be called out explicitly (and perhaps other states too).

This is similar to how `blockchain.scripthash.subscribe` works. Clients can already subscribe to any address (scripthash) regardless if it has existing history or not. I don't see a fundamental reason why it should be different here. I think clients should be able to subscribe to non-existent outpoints and get a notification when they get funded.

I think from the text as-is, it is fairly clear that the case where the funding output does not exist should be handled (see e.g. "Result" section "The output is a dictionary, containing 0, 1, or 3 of the following items" (where 0 corresponds to this case)).
You are right that it is not explicitly pointed out that even in that case the server should save the subscription, however I would argue that given this method is a "subscription", the default assumption should be that a subscription is made. The cases added just recently re deeply mined txs (where the status ~cannot change anymore) are the exception, not the norm.

Reply:
The reason I would expect it differently is that scripts exist at the client. Only now I realized there's the case when the client broadcasts transaction and subscribes to the outpoint immediately after - before the server gets the chance to receive and process that transaction. Maybe others could make the same mistake?

I believe simple phrase "Note that the server MUST *subscribe* the outpoint even if it doesn't exist yet." Has pretty good benefit/cost ratio.

Reply:
> I believe simple phrase "Note that the server MUST _subscribe_ the outpoint even if it doesn't exist yet." Has pretty good benefit/cost ratio.

I must chime in here -- although I haven't followed the discussion fully -- I really do think that is a good idea. And, in some ways, it also mimics how address subscriptions work (where you can subscribe to "nonsense" at your heart's content, up to a server-defined internal limit).  

Anyway I think this is a good idea to specify in the protocol that servers ***must*** honor subscriptions for outpoints that "don't exist yet", with perhaps an implementation-defined limit on such subs and/or a timeout on such thing, etc.

---

Related -- Sorry if this has been mentioned before -- it's hard to go back and read the pages of text in this thread, but: 

Perhaps also, servers should probably keep old outpoint subs around, even after they are spent in a confirmed tx (and thus under normal circumstances would not generate new events, so they would be ripe for deletion). Why? To guard against some sort of pathological reorg scenario leading to lost notifications.  Maybe is should keep them around using implementation-defined heuristics such as maybe some "safe" confirmed depth has elapsed.  This way even a pathological/rare reorg circumstance wouldn't lead to a situation where subs are "lost" (even though outpoint events would have fired had they not been). This last point doesn't necessarily need to be specified in the protocol though, perhaps.

Reply:
> I believe simple phrase "Note that the server MUST subscribe the outpoint even if it doesn't exist yet." Has pretty good benefit/cost ratio.

Ok, added in https://github.com/spesmilo/electrumx/pull/90/commits/8b08ebc8ea82b8f5e289e577a245a3c0af13b6b6

> Perhaps also, servers should probably keep old outpoint subs around, even after they are spent in a confirmed tx (and thus under normal circumstances would not generate new events, so they would be ripe for deletion). Why? To guard against some sort of pathological reorg scenario leading to lost notifications. Maybe is should keep them around using implementation-defined heuristics such as maybe some "safe" confirmed depth has elapsed. 

Something like this was already discussed: we were more concerned about the other direction: the server never deleting subs. I agree the server should keep subs around for a long time and only potentially clean them up if both txs are deeply mined. See https://github.com/spesmilo/electrumx/pull/90/commits/4e25d7cc2501244906a77ac83b509ef7207dc3ba

Reply:
> perhaps an implementation-defined limit on such subs

Hmm, good point though it should be advertised in response (`exceeds_limit: true`). Silently missing notifications could be quite bad.

Reply:
Yeah, good point. It always irked me that that can happen, as far as the protocol is concerned, and the client may be left unaware of that fact.

I think on my Fulcrum server I return a JSON-RPC "error" response in this case on scripthash subs, in protocol v1.4, if the limit is exceeded (which may be non-standard for this protocol, but it bothered me so bad I wanted it to explicitly do that...). 

So yeah an alternative could be an app-level JSON-RPC error. Although a non-error response with a key like that is also fine...

Reply:
I don't think it's worth to add error codes and limits specific to number of outpoint subscriptions in the protocol documentation here. Again, the same situation applies to scripthash subs where it is left for the implementation to handle.

There used to be a configurable limit in ElectrumX for max number of address subs for a given session, over which the server sent an error and force-disconnected the session. Now ElectrumX instead has generic session cost accounting with a soft and a hard limit (when soft limit is reached, session is throttled/ratelimited; if hard limit is reached, an error is sent and then the session is disconnected).

Reply:
> he same situation applies to scripthash subs where it is left for the implementation to handle

Sounds like that could also be improved by adding the key there as well?

> if hard limit is reached, an error is sent and then the session is disconnected).

That works although feels a bit drastic.

Reply:
> I don't think it's worth to add error codes and limits specific to number of outpoint subscriptions in the protocol documentation here. Again, the same situation applies to scripthash subs where it is left for the implementation to handle.

Ok, that's fair. I can see why perhaps specifying that in the protocol constrains servers from doing clever things.